Improve intro workout in react (#1500)
enkidevs/curriculum: javascript/react/intro-react/why-use-react.md

  - introduction

---
# Why use React?

---
## Content

**React** is considered a great choice for building the **UI** of web applications - be it simple *SPAs*[1] or large projects with numerous authors due to the **self-containing** nature of its *components*.

In addition to the previously stated reasons developers are eager to adopt this library, **React** comes with many more interesting features such as:

> - great *performance* as a result of the **Virtual DOM** diff algorithm

Reviewer: Add footnote?

Author: Added the following footnote

The Virtual DOM (or VDOM) is nothing more but a programming concept in which a "virtual" representation of the UI is kept in memory. It is synced with the "real" DOM by using a library, which in our case is the `ReactDOM` library. The Virtual DOM provides increased performance because only the parts of the app that have been changed are re-rendered, instead of re-rendering the whole application.
```jsx
<div className="first" />

<div className="second"/>
```
In this case, React know to only modify the `className` on the underlying DOM node. For more information on the **diff algorithm** that React uses make sure to check out [this article](https://reactjs.org/docs/reconciliation.html).

> - **React**'s syntax is simple. Unlike other popular JS frameworks with specific syntax, **React** makes use of pure JS

Reviewer: We probably shouldn't compare it to other frameworks, particularly in this case where "simplicity" is a subjective trait.

We should instead talk about how the syntax is ["declarative"](https://stackoverflow.com/questions/129628/what-is-declarative-programming)

Author: changed bullet point to:

- **React**'s syntax is declarative[3]

[3:Declarative Syntax]
This means that you write your code such that it describes what you want to do, not how you want to do it. Figuring out how is left to the compiler.

- easier debugging with **out-of-the-box tools** such as `React Chrome Extension`

> - **React** is good for *Search Engine Optimization* when employing server-side rendering

Reviewer: Hmmmm not really a benefit. In fact, SEO is more difficult because of React, not vice-versa. We should remove this.

> - knowing **React** allows you to easily switch to **React Native** to build applications

Reviewer: ```suggestion
- knowing **React** allows you to easily switch to **React Native** to build native mobile applications
```

**Single-Page Applications** (SPAs) are Web apps that load a single `HTML` page and dynamically update that page as the user interacts with the app.

[2:VDOM]

> The Virtual DOM (or VDOM) is nothing more but a programming concept in which a "virtual" representation of the UI is kept in memory. It is synced with the "real" DOM by using a library, which in our case is the `ReactDOM` library. The Virtual DOM provides increased performance because only the parts of the app that have been changed are re-rendered, instead of re-rendering the whole application.

Reviewer: ```suggestion
The Virtual DOM (or VDOM) is a programming concept in which a "virtual" representation of the DOM is kept in memory. It is synced with the "real" DOM by using a library, which in React's case is the `ReactDOM` library. The Virtual DOM provides increased performance because it reduced the need for expensive interactions with the actual DOM. Most data changes in React stay in the virtual, in-memory DOM and only propagate to the actual DOM if necessary.
```

> The Virtual DOM (or VDOM) is a programming concept in which a "virtual" representation of the DOM is kept in memory. It is synced with the "real" DOM by using a library, which in React's case is the `ReactDOM` library. The Virtual DOM provides increased performance because it reduced the need for expensive interactions with the actual DOM. Most data changes in React stay in the virtual, in-memory DOM and only propagate to the actual DOM if necessary.

Reviewer: ```suggestion
The Virtual DOM (or VDOM) is a programming concept in which a "virtual" representation of the DOM is kept in memory. It is synced with the "real" DOM by using a library, which in React's case is the `ReactDOM` library. The Virtual DOM provides increased performance because it reduces the need for expensive interactions with the actual DOM. Most data changes in React stay in the virtual, in-memory DOM and only propagate to the actual DOM if necessary.
```

```jsx
<div className="first" />

<div className="second"/>
```

> In this case, React know to only modify the `className` on the underlying DOM node. For more information on the **diff algorithm** that React uses make sure to check out [this article](https://reactjs.org/docs/reconciliation.html).

Reviewer: ```suggestion
In this case, React knows to only modify the `className` on the underlying DOM node. For more information on the **diff algorithm** that React uses to make sure to check out [this article](https://reactjs.org/docs/reconciliation.html).
```

> In this case, React knows to only modify the `className` on the underlying DOM node. For more information on the **diff algorithm** that React uses to make sure to check out [this article](https://reactjs.org/docs/reconciliation.html).

Reviewer: ```suggestion
In this case, React knows to only modify the `className` on the underlying DOM node. For more information on the **diff algorithm** that React uses, make sure to check out [this article](https://reactjs.org/docs/reconciliation.html).
```

[3:Declarative Syntax]

> This means that you write your code such that it describes what you want to do, not how you want to do it. Figuring out how is left to the compiler.

Reviewer: ```suggestion
This means that you write your code such that it describes what you want to do, not how you want to do it. Figuring out how is an implementation detail of the runtime system.
```
